packages/graph/hash_graph/lib/graph/src/datastore/mod.rs

```diff
@@ -146,19 +148,43 @@ impl fmt::Display for DatabaseConnectionInfo {
 /// Describes the API of a Datastore implementation
 #[async_trait]
 trait Datastore {
-    async fn create_data_type() -> Result<(), DatastoreError>;
-
-    async fn get_data_type() -> Result<(), DatastoreError>;
+    /// Creates a new [`DataType`].
+    ///
+    /// # Errors:
+    ///
+    /// If the account referred to by `created_by` does not exist.
+    async fn create_data_type(
+        &self,
+        data_type: DataType,
+        created_by: AccountId,
+    ) -> Result<Qualified<DataType>, DatastoreError>;
+
+    /// Get an existing [`DataType`] by an [`Identifier`].
+    ///
+    /// # Errors
+    ///
+    /// If the [`DataType`] doesn't exist.
+    async fn get_data_type(&self, id: Identifier) -> Result<Qualified<DataType>, DatastoreError>;
 
-    async fn get_data_many() -> Result<(), DatastoreError>;
+    async fn get_data_type_many() -> Result<(), DatastoreError>;
 
-    async fn update_data_type() -> Result<(), DatastoreError>;
+    /// Update the definition of an exisitng [`DataType`].
+    ///
+    /// # Errors
+    ///
+    /// If the [`DataType`] doesn't exist.
```

Comment:
Not requesting a change but I wonder if we should consider `update` operations that create if it does not exist.

I find it really annoying to have to do:
Read to check and then
Create or Update

Reply:
I wonder if we can leverage Postgres' [`ON CONFLICT`](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT) upserts to accomplish this behavior in SQL

Reply:
My instinct is that it would be better to have a separate `updateOrCreate` method that encapsulates this logic and the calls to `update` and `create`, as it could be confusing for this to be the default behaviour.

A caller of the method may be unaware they are passing the wrong id, think they're correctly updating an existing type, but instead they've ended up creating a new entity in the graph.

Reply:
That's also a valid point Ben. The difference in signature with the IDs might make this a little questionable. I think it'll be a lot harder to accidentally push a mistyped URI though.

Lets discuss more later :) As I said, definitely no need to address in this PR